crates/shackle-ls/src/handlers/rename_symbol.rs

```diff
@@ -0,0 +1,154 @@
+use lsp_server::ResponseError;
+use lsp_types::{request::Rename, RenameParams, TextEdit, WorkspaceEdit};
+use shackle::{
+	db::CompilerDatabase,
+	file::ModelRef,
+	hir::{
+		db::Hir,
+		ids::{LocalEntityRef, NodeRef, PatternRef},
+		source::{find_node, Point},
+	},
+	syntax::db::SourceParser,
+	utils,
+};
+use std::collections::HashMap;
+
+use crate::{db::LanguageServerContext, dispatch::RequestHandler, utils::node_ref_to_location};
+
+#[derive(Debug)]
+pub struct RenameHandler;
+
+pub struct SymbolHandlerData {
+	model_ref: ModelRef,
+	cursor_pos: Point,
+	new_name: String,
+}
+
+impl RequestHandler<Rename, SymbolHandlerData> for RenameHandler {
+	fn prepare(
+		db: &mut impl LanguageServerContext,
+		params: RenameParams,
+	) -> Result<SymbolHandlerData, ResponseError> {
+		// Verify if the new name is valid
+		let mut attempted_new_name = params.new_name.chars();
+
+		// identifiers can consist of anything which is not in these chars
+		let cursor_pos = Point {
+			row: params.text_document_position.position.line as usize,
+			column: params.text_document_position.position.character as usize,
+		};
+
+		// cannot include single quotes
+		if attempted_new_name.any(|ch| ch == '\'') {
+			return Err(ResponseError {
+				code: -32602,
```

Comment:
It would probably be better to use the constant `lsp_server::ErrorCode::InvalidRequest as i32` rather than using the number directly